### DISCUSSION TITLE
Trouble with development environment

### DISCUSSION THREAD

**Discussion by tibortakacs** [tox-dev/tox#2833] [Q&A] [answered]
Thank you for the hard work on the latest tox. I like the new, simplified plugin model. I believe the documentation does not cover all use cases though, and I am busy with a new plugin to support a special installation step. Please, bear with me if this is not right forum to ask, but can I ask some guidance regarding the following plugin tasks:

* I would like to add a new `proxy` core config (`tox_add_env_config`). Later, during the installation (`tox_on_install`), I want to use a special install command based on the value of `proxy`. So, somewhere in the call chain, I would like to overwrite the `install_command` config. (I do understand that I could just add a `install_command` to `tox.ini`, I would like to dynamically decide what the right install command is.) That said, **is there any way to set or overwrite a config value in a plugin?** Is only way that I create a new environment?
* How does the new environment registration work? How is the right environment selected if there are multiple one? Or, if I register my environment, does overwrite the existing one?

Thank you in advance.

**Comment by gaborbernat** [comment, marked as answer]
> * I would like to add a new `proxy` core config (`tox_add_env_config`). Later, during the installation (`tox_on_install`), I want to use a special install command based on the value of `proxy`. So, somewhere in the call chain, I would like to overwrite the `install_command` config. (I do understand that I could just add a `install_command` to `tox.ini`, I would like to dynamically decide what the right install command is.) That said, **is there any way to set or overwrite a config value in a plugin?** Is only way that I create a new environment?

From within a plugin you can insert a memory loader earlier into the source chain as done here https://github.com/tox-dev/tox/blob/main/src/tox/session/cmd/exec_.py#L31-L37


> * How does the new environment registration work? How is the right environment selected if there are multiple one? Or, if I register my environment, does overwrite the existing one?

https://github.com/tox-dev/tox/blob/main/src/tox/session/env_select.py#L309-L330

**Reply by tibortakacs** [reply]
Brilliant, thank you. Should I insert the loader into a new environment, or can I just do it in the existing `ToxEnv` in the on_install or on_option hook?

**Reply by gaborbernat** [reply]
for `tox config` to work you should do it within https://tox.wiki/en/latest/plugins.html#tox.plugin.spec.tox_add_env_config

**Comment by tibortakacs** [comment]
Thank you for the answer, it works well.
Is it possible to apply similar approach to first read the predefined option value, and then using the Memory Loader, update it? It looks like for me that once and option is read, the options are sealed, and the new loaders do not have any effect.

**Reply by gaborbernat** [reply]
As the example demonstrates, to read and change a value, you must change it inline (see env_list). We don't allow currently reloading a value from new loaders. See https://github.com/tox-dev/tox/discussions/2882

**Reply by tibortakacs** [reply]
Thank you. This example is great to add new env dynamically. I am still searching for how this approach could help to modify an existing option (e.g. the install_command) based on some configuration value? For example, if the custom parameter 'nodep'  is set, it extends the install command with - - no-dep.

**Reply by gaborbernat** [reply]
I don't recommending amending the install command, I'd suggest set PIP_NO_DEP env-var instead on setenv 👍  You can read it and then set it 👍 on the object you got.

**Reply by tibortakacs** [reply]
Great suggestion, thank you.
I also want to intercept the PIP_INDEX_URL: if a custom tox cli parameter is used, and want to set PIP_INDEX_URL to a certain value. Additionally, if it is already set, I want to read its value first. Is that possible?

**Reply by gaborbernat** [reply]
Check the test suite there are examples for that https://github.com/tox-dev/tox/blob/main/tests/plugin/test_plugin.py#L186
